rasa/shared/core/domain.py

```diff
@@ -256,61 +259,127 @@ def merge(self, domain: Optional["Domain"], override: bool = False) -> "Domain":
+
+        domain_dict = domain2
+        combined = domain1
 
         if override:
             config = domain_dict["config"]
             for key, val in config.items():
                 combined["config"][key] = val
 
-        if override or self.session_config == SessionConfig.default():
+        if override or domain2.get("session_config"):
             combined[SESSION_CONFIG_KEY] = domain_dict[SESSION_CONFIG_KEY]
 
-        combined[KEY_INTENTS] = merge_lists_of_dicts(
-            combined[KEY_INTENTS], domain_dict[KEY_INTENTS], override
-        )
+        if combined[KEY_INTENTS] or domain_dict[KEY_INTENTS]:
+            combined[KEY_INTENTS] = self.merge_lists_of_dicts(
+                combined[KEY_INTENTS], domain_dict[KEY_INTENTS], override
+            )
 
         # remove existing forms from new actions
-        for form in combined[KEY_FORMS]:
-            if form in domain_dict[KEY_ACTIONS]:
+        for form in combined.get(KEY_FORMS, []):
+            if form in domain_dict.get(KEY_ACTIONS, []):
                 domain_dict[KEY_ACTIONS].remove(form)
 
         for key in [KEY_ENTITIES, KEY_ACTIONS, KEY_E2E_ACTIONS]:
-            combined[key] = merge_lists(combined[key], domain_dict[key])
+            combined[key] = self.merge_lists(
+                combined.get(key, []), domain_dict.get(key, [])
+            )
 
         for key in [KEY_FORMS, KEY_RESPONSES, KEY_SLOTS]:
-            combined[key] = merge_dicts(combined[key], domain_dict[key], override)
+            combined[key] = self.merge_dicts(
+                combined.get(key, {}), domain_dict.get(key, {}), override
+            )
 
-        return self.__class__.from_dict(combined)
+        return combined
+
+    @staticmethod
+    def merge_dicts(
+        tempDict1: Dict[Text, Any],
+        tempDict2: Dict[Text, Any],
+        override_existing_values: bool = False,
+    ) -> Dict[Text, Any]:
+        """Merges 2 dicts."""
+        if override_existing_values:
+            merged_dicts, b = tempDict1.copy(), tempDict2.copy()
+
+        else:
+            merged_dicts, b = tempDict2.copy(), tempDict1.copy()
+        merged_dicts.update(b)
+        return merged_dicts
+
+    @staticmethod
+    def merge_lists(list1: List[Any], list2: List[Any]) -> List[Any]:
+        """Merges 2 lists."""
+        return sorted(list(set(list1 + list2)))
+
+    @staticmethod
+    def merge_lists_of_dicts(
+        dict_list1: List[Dict],
+        dict_list2: List[Dict],
+        override_existing_values: bool = False,
+    ) -> List[Dict]:
+        """Merges 2 dict lists."""
+        dict1 = {
+            (sorted(list(i.keys()))[0] if isinstance(i, dict) else i): i
+            for i in dict_list1
+        }
+        dict2 = {
+            (sorted(list(i.keys()))[0] if isinstance(i, dict) else i): i
+            for i in dict_list2
+        }
+        merged_dicts = Domain.merge_dicts(dict1, dict2, override_existing_values)
+        return list(merged_dicts.values())
```

Comment:
I don't see any duplicates extraction when needs to be done when merging the domain.
Here is where it currently is: https://github.com/RasaHQ/rasa/blob/main/rasa/shared/core/domain.py#L291

Reply:
Hey @alwx the method you've linked to is on main, but this PR is merging into 2.8.x, which doesn't actually have the `extract_duplicates` method that you've linked to.
We're planning to eventually merge this into `main` (I've raised a [ticket](https://github.com/RasaHQ/rasa/issues/10570) for this), so will add a note there about being careful to observe any changes in domain logic between 2.8.x and 3.0.x.

Reply:
Oh, that‘s my bad then, haven’t noticed that. Thanks for explanation!